firewalldb/actions_sql.go

```diff
@@ -0,0 +1,418 @@
+		}
+		intent = sql.NullString{
+			String: req.Intent,
+			Valid:  req.Intent != "",
+		}
+	)
+
+	err := s.db.ExecTx(ctx, &writeTxOpts, func(db SQLQueries) error {
+		// Do best effort to see if this action is linked to a session,
+		// and/or an action or none.
+		var (
+			sessionID sql.NullInt64
+			accountID sql.NullInt64
+		)
+
+		// First check session DB.
+		var sessErr error
+		req.SessionID.WhenSome(func(id session.ID) {
+			sessID, err := db.GetSessionIDByAlias(ctx, id[:])
+			if errors.Is(err, sql.ErrNoRows) {
+				sessErr = session.ErrSessionNotFound
+				return
+			} else if err != nil {
+				sessErr = err
+				return
+			}
+
+			sessionID = sqldb.SQLInt64(sessID)
+		})
+		if sessErr != nil {
+			return sessErr
+		}
+
+		// If an account ID was provided, then it must exist in our DB.
+		var getAcctErr error
+		req.AccountID.WhenSome(func(alias accounts.AccountID) {
+			aliasInt, err := alias.ToInt64()
+			if err != nil {
+				getAcctErr = err
+				return
+			}
+
+			acctID, err := db.GetAccountIDByAlias(ctx, aliasInt)
+			if errors.Is(err, sql.ErrNoRows) {
+				getAcctErr = accounts.ErrAccNotFound
+				return
+			} else if err != nil {
+				getAcctErr = err
+				return
+			}
+
+			accountID = sqldb.SQLInt64(acctID)
+		})
+		if getAcctErr != nil {
+			return getAcctErr
+		}
+
+		var macID []byte
+		req.MacaroonIdentifier.WhenSome(func(id [4]byte) {
+			macID = id[:]
+		})
+
+		id, err := db.InsertAction(ctx, sqlc.InsertActionParams{
+			SessionID:          sessionID,
+			AccountID:          accountID,
+			ActorName:          actor,
+			MacaroonIdentifier: macID,
+			FeatureName:        feature,
+			ActionTrigger:      trigger,
+			Intent:             intent,
+			StructuredJsonData: []byte(req.StructuredJsonData),
+			RpcMethod:          req.RPCMethod,
+			RpcParamsJson:      req.RPCParamsJson,
+			CreatedAt:          s.clock.Now().UTC(),
+			ActionState:        int16(ActionStateInit),
+		})
+		if err != nil {
+			return err
+		}
+
+		locator = sqlActionLocator{
+			id: id,
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &locator, nil
+}
+
+// SetActionState finds the action specified by the ActionLocator and sets its
+// state to the given state.
+//
+// NOTE: This is a part of the ActionDB interface.
+func (s *SQLDB) SetActionState(ctx context.Context, al ActionLocator,
+	state ActionState, errReason string) error {
+
+	if errReason != "" && state != ActionStateError {
+		return fmt.Errorf("error reason should only be set for " +
+			"ActionStateError")
+	}
+
+	locator, ok := al.(*sqlActionLocator)
+	if !ok {
+		return fmt.Errorf("expected sqlActionLocator, got %T", al)
+	}
+
+	var writeTxOpts db.QueriesTxOptions
+	return s.db.ExecTx(ctx, &writeTxOpts, func(db SQLQueries) error {
+		return db.SetActionState(ctx, sqlc.SetActionStateParams{
+			ID:          locator.id,
+			ActionState: int16(state),
+			ErrorReason: sql.NullString{
+				String: errReason,
+				Valid:  errReason != "",
```

Comment:
nit: we could set it `true` here, since that must be the case

Reply:
not true right? it is defs possible for SetAction to be called with an empty `errReason`